routes/artifacts.js

```diff
@@ -0,0 +1,597 @@
+    "",
+    "**Artifact immutability**, generally speaking you cannot overwrite an",
+    "artifact when created. But if you repeat the request with the same",
+    "properties the request will succeed as the operation is idempotent.",
+    "This is useful if you need to refresh a signed URL while uploading.",
+    "Do not abuse this to overwrite artifacts created by another entity!",
+    "Such as worker-host overwriting artifact created by worker-code.",
+    "",
+    "As a special case the `url` property on _reference artifacts_ can be",
+    "updated. You should only use this to update the `url` property for",
+    "reference artifacts your process has created."
+  ].join('\n')
+}, async function(req ,res) {
+  // Validate parameters
+  if (!api.checkParams(req, res)) {
+    return;
+  }
+
+  var taskId      = req.params.taskId;
+  var runId       = parseInt(req.params.runId);
+  var name        = req.params.name;
+  var input       = req.body;
+  var storageType = input.storageType;
+  var contentType = input.contentType || 'application/json';
+
+  // Find expiration date
+  var expires = new Date(input.expires);
+
+  // Validate expires it is in the future
+  var past = new Date();
+  past.setMinutes(past.getMinutes() - 15);
+  if (expires.getTime() < past.getTime()) {
+    return res.status(400).json({
+      message:  "Expires must be in the future",
+      error: {
+        now:      new Date().toJSON(),
+        expires:  expires.toJSON()
+      }
+    });
+  }
+
+  // Load Task entity
+  var task = await this.Task.load({
+    taskId:       taskId
+  }, true);
+
+  // Handle cases where the task doesn't exist
+  if (!task) {
+    return res.status(404).json({
+      message: "Task not found"
+    });
+  }
+
+  // Check presence of the run
+  var run = task.runs[runId];
+  if (!run) {
+    return res.status(404).json({
+      message: "Run not found"
+    });
+  }
+
+  // Get workerGroup and workerId
+  var workerGroup = run.workerGroup;
+  var workerId    = run.workerId;
+
+  // Authenticate request by providing parameters
+  if(!req.satisfies({
+    workerGroup:  workerGroup,
+    workerId:     workerId,
+    name:         name
+  })) {
+    return;
+  }
+
+  // Validate expires <= task.expires
+  if (expires.getTime() > task.expires.getTime()) {
+    return res.status(400).json({
+      messages: "Artifact expires before the task (task.expires < expires)",
+      error: {
+        taskExpires:      task.expires.toJSON(),
+        expires:          expires.toJSON()
+      }
+    });
+  }
+
+  // Ensure that the run is running
+  if (run.state !== 'running') {
+    return res.status(409).json({
+      message:    "The given is not running",
+      error: {
+        status:   task.status(),
+        runId:    runId
+      }
+    });
+  }
+
+  // Construct details for different storage types
+  var isPublic = /^public\//.test(name);
+  var details  = {};
+  if (storageType === 's3') {
```

Comment:
Switch statement didn't get shorter... and I generally try to keep indentation limited..
But in this case I seems to make sense...